### PR TITLE
Issue #28, Caching of Non-Temporaries in the IJ-Plane

### PR DESCRIPTION
**Pull request by twicki**
## Brief

Addition of PassSetNonTempCaches that decides, which non-temporary variables we should write to a scarchpad based on the metric implemented in  PassDataLocalityMetric . It adds filler / flush stages if need be at the beginning and the end of any given multistage the pass is run on. 
In order to run with this pass enabled, the flag `-fcache-non-temp-fields` can be set.

Small adaptation to the PassDataLocalityMetric to support the updated needs.

### CODE REVIEW

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+  accessIDaccessMetric(int id, int gain) : accessID(id), dataLocalityGain(gain) {}
+  int accessID;
+  int dataLocalityGain;
+  bool operator<(const accessIDaccessMetric& rhs) {
+    return dataLocalityGain < rhs.dataLocalityGain;
+  }
+};
+
+struct nameToImprovementMetric {
+  nameToImprovementMetric(std::string name_, Cache cache_, int metric)
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: This is not needed in C++11 see http://en.cppreference.com/w/cpp/language/aggregate_initialization

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+        }
+      }
+      return true;
+  }
+
+  MultiStage* multiStagePrt_;
+  HardwareConfig config_;
+  StencilInstantiation* instantiation_;
+
+  std::unordered_map<int, int> accessIDtoDataLocality_;
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: accessID**T**oDataLocality_

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+  std::unordered_map<int, int> oldIDtoNewID;
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: which IDs do you refer to here? Maybe call it `oldAccessIDToNewAccessID_`

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+            stmtAccessesPair->getCallerAccesses()->getWriteAccesses().find(id);
+        if(wirteAccessIterator != stmtAccessesPair->getCallerAccesses()->getWriteAccesses().end()) {
+          return false;
+        }
+      }
+    }
+    return false;
+  }
+
+  bool checkReadOnlyAccess(int id) {
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: It would be better to stick to the convention of calling it `AccessID` instead of just `id`.

The formatting is really broken in this function

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+  bool checkReadBeforeWrite(int id) {
+    for(auto stageItGlob = multiStagePrt_->getStages().begin();
+        stageItGlob != multiStagePrt_->getStages().end(); ++stageItGlob) {
+      DoMethod& doMethod = (**stageItGlob).getSingleDoMethod();
+      for(int stmtIndex = 0; stmtIndex < doMethod.getStatementAccessesPairs().size(); ++stmtIndex) {
+        const std::shared_ptr<StatementAccessesPair>& stmtAccessesPair =
+            doMethod.getStatementAccessesPairs()[stmtIndex];
+
+        // Find first if this statement has a read
+        auto readAccessIterator = stmtAccessesPair->getCallerAccesses()->getReadAccesses().find(id);
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: It think for sake of readability it is worth storing the read accesses in a local variable to avoid the long chains `stmtAccessesPair->getCallerAccesses()->getReadAccesses()` (you access it several times)

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+    assignmentStage->update();
+
+    return assignmentStage;
+  }
+
+  ///@brief add the assignment operator of two unique id's to a given domethod
+  void addAssignmentToDoMethod(std::unique_ptr<DoMethod>& domethod, int assignmentID,
+                               int assigneeID) {
+    // Create the StatementAccessPair of the assignment with the new and old variables
+    std::shared_ptr<FieldAccessExpr> fa_assignee =
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: It is very hard to follow these few lines...  It would strongly recommend using `auto` in this case as the syntactic noise from `shared_ptr` here is crazy.

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+  /// We always fill to the extent of the given multistage and we only add one stage to fill all the
+  /// variables to reduce synchronisation overhead
+  void addFillerStages() {
+    for(int i = 0; i < std::min((int)sortedAccesses_.size(), config_.SMemMaxFields); ++i) {
+      int oldID = sortedAccesses_[i].accessID;
+
+      // Create new temporary field and register in the instantiation
+      int newID = instantiation_->nextUID();
+
+      instantiation_->setAccessIDNamePairOfField(newID, dawn::format("__tmp_cache_%i", i), true);
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: `instantiation_->setAccessIDNamePairOfField(newID, "__tmp_cache_" + std::to_string(i), true);` [this an order of magnitude faster ;)]

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+  }
+
+  /// @brief for each chached variable, check if we need a filler-stage (if we find a
+  /// read-before-write) and add it if necessary, renames all occurances and sets the cache for the
+  /// temporary variable newly created
+  ///
+  /// We always fill to the extent of the given multistage and we only add one stage to fill all the
+  /// variables to reduce synchronisation overhead
+  void addFillerStages() {
+    for(int i = 0; i < std::min((int)sortedAccesses_.size(), config_.SMemMaxFields); ++i) {
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: It would be better to precompute the upper bound of the loop as a general rule of thumb.

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+      multiStagePrt_->renameAllOccurrences(oldID, newID);
+
+      oldIDtoNewID.emplace(oldID, newID);
+      Cache& cache = multiStagePrt_->setCache(Cache::IJ, Cache::local, newID);
+      originalNameToCache_.emplace_back(instantiation_->getOriginalNameFromAccessID(oldID), cache,
+                                        accessIDtoDataLocality_.find(oldID)->second);
+    }
+
+    // Create the cache-filler stage
+    std::vector<int> oldIDs;
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: It would advocate to use `oldAccessIDs`

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+  void process() {
+    computeOptimalFields();
+    addFillerStages();
+    addFlushStages();
+  }
+
+  std::vector<nameToImprovementMetric> getoriginalNameToCache() { return originalNameToCache_; }
+
+private:
+  /// @brief use the data locality metric to rank the fields in a stencil based on how much they
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: For user of the documentation it would be good to stick to one convention and always start the brief statement of doxygen with a capital letter

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+    std::vector<int> oldIDs;
+    std::vector<int> newIDs;
+    for(auto& idPair : oldIDtoNewID) {
+      if(!checkReadOnlyAccess(idPair.second)) {
+        oldIDs.push_back(idPair.first);
+        newIDs.push_back(idPair.second);
+      }
+    }
+
+    if(oldIDs.size() > 0) {
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: How about `if(!oldIDs.empty())`. I think this a relays the intent slightly better.

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+#include "dawn/Optimizer/StencilInstantiation.h"
+#include "dawn/SIR/ASTExpr.h"
+#include "dawn/Support/Unreachable.h"
+#include <iostream>
+#include <set>
+#include <vector>
+
+namespace dawn {
+
+struct accessIDaccessMetric {
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: Classes should start with a capital letter.

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+
+  /// @brief entry method for the pass: processes a given multistage and applies all changes
+  /// required
+  void process() {
+    computeOptimalFields();
+    addFillerStages();
+    addFlushStages();
+  }
+
+  std::vector<nameToImprovementMetric> getoriginalNameToCache() { return originalNameToCache_; }
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: You should almost never return by **value** a member of a class. It is almost always beneficial to return a constant reference and you call this function as an in (line 298) a loop which could be a disaster (In this case you only waste one copy as containers in range based for-loops are guaranteed to be evaluated only once - https://stackoverflow.com/questions/10821756/is-the-ranged-based-for-loop-beneficial-to-performance)

Also the name scheme is not correct it should be get**O**riginalNameToCache

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 14]
```diff
@@ -0,0 +1,325 @@
...
+struct nameToImprovementMetric {
+  nameToImprovementMetric(std::string name_, Cache cache_, int metric)
+      : name(name_), cache(cache_), dataLocalityImprovement(metric) {}
+  std::string name;
+  Cache cache;
+  int dataLocalityImprovement;
+};
+
+/// @brief The GlobalFieldCacher class handles the caching for a given Multistage
+/// @param[in, out]  msprt   Pointer to the multistage to handle
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: This comment should be part of the constructor?

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 15]
```diff
@@ -0,0 +1,325 @@
...
+  }
+
+  std::vector<nameToImprovementMetric> getoriginalNameToCache() { return originalNameToCache_; }
+
+private:
+  /// @brief use the data locality metric to rank the fields in a stencil based on how much they
+  /// would benefit from caching
+  void computeOptimalFields() {
+    auto dataLocality =
+        computeReadWriteAccessesMetricIndividually(instantiation_, *multiStagePrt_, config_);
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: How about naming this function `computeReadWriteAccessesMetricPerAccessID`? It is not immediately clear what individually means in this context (does it refer to the metric or does it mean individually for each field).

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 16]
```diff
@@ -0,0 +1,325 @@
...
+
+    for(const auto& stagePtr : multiStagePrt_->getStages()) {
+      for(const Field& field : stagePtr->getFields()) {
+        auto findField = accessIDtoDataLocality_.find(field.AccessID);
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: `findField` is not a very good name here, I don't think this is even a field? Given that this is actually an `int` I would assume from the name of the map this is the value of the metric, no?

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 17]
```diff
@@ -0,0 +1,325 @@
...
+
+        // We already checked this field
+        if(findField != accessIDtoDataLocality_.end())
+          continue;
+
+        // Only ij-cache fields that are not horizontally pointwise and are vertically pointwise
+        if(!(field.Extent.isVerticalPointwise()) || (field.Extent.isHorizontalPointwise()))
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: parenthesis are not needed here operator **!** takes precedence over **||**

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 18]
```diff
@@ -0,0 +1,325 @@
...
+          continue;
+
+        // This is caching non-temporary fields
+        if(instantiation_->isTemporaryField(field.AccessID))
+          continue;
+
+        int cachedreadwrites = dataLocality.find(field.AccessID)->second.total();
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: cached**R**ead**And**Writes or something like that

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 19]
```diff
@@ -0,0 +1,325 @@
...
+        // We need one more write to flush the cache back to the variable
+        if(!checkReadOnlyAccess(field.AccessID))
+          cachedreadwrites--;
+
+        if(cachedreadwrites > 0)
+          accessIDtoDataLocality_.emplace(field.AccessID, cachedreadwrites);
+      }
+    }
+
+    for(auto& performanceMesure : accessIDtoDataLocality_) {
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: What does `performanceMesure` mean here? Isn't this just a pair of AccessIDs to the value of the metric? How about naming this `AccessIDMetricPair`?

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 20]
```diff
@@ -0,0 +1,325 @@
...
+          if(wirteAccessIterator != stmtAccessesPair->getCallerAccesses()->getWriteAccesses().end()) {
+            return false;
+          }
+        }
+      }
+      return true;
+  }
+
+  MultiStage* multiStagePrt_;
+  HardwareConfig config_;
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: This is never initialized and just relies on default initialization. We should open an issue that the `HardwareConfig` gets properly passed around in the Passes. One way would be to extend the `Pass::run()` to take an additional argument i.e the `HardwareConfig`

File: src/dawn/Optimizer/PassSetNonTempCaches.h [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+#ifndef DAWN_OPTIMIZER_PASSNONTEMPCACHES_H
+#define DAWN_OPTIMIZER_PASSNONTEMPCACHES_H
+
+#include "dawn/Optimizer/Pass.h"
+
+namespace dawn {
+
+/// @brief This Pass optimizes data locality on a per-multistage basis
+///
+/// For all the non-temporary fields that we access in any given multistage, we check how high the
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.h]: For all the non-temporary, vertically-pointwise fields ... Then you can skip the next sentence

File: src/dawn/Optimizer/PassSetNonTempCaches.h [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+/// performance gain of using a local ij-cache woud be. This is only beneficial if the fields are
+/// horizontally non-pointwise but vertically pointwise.
+/// If is is beneficial, we cache the fieldswith the most performance gain up to HardwareConfig's
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.h]: fields with

File: src/dawn/Optimizer/PassSetNonTempCaches.h [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+/// Maximum number of fields concurrently in shared memory (defined in PassDataLocalityMetric.h)
+///
+/// The caching invokes a renaming of all the occurances of any given field we cache and adds
+/// potentially filler stages and definitly flush stages to the multistage thus changing the ISIR
```
Comment (thfabian) [inline src/dawn/Optimizer/PassSetNonTempCaches.h]: definitely

File: src/dawn/Optimizer/MultiStage.h [hunk 24]
```diff
@@ -118,6 +118,9 @@ class MultiStage {
   /// @brief Get the caches
   std::unordered_map<int, Cache>& getCaches() { return caches_; }
   const std::unordered_map<int, Cache>& getCaches() const { return caches_; }
+
+  /// @brief Rename all the occurances in the multi-stage
+  void renameAllOccurrences(int oldAccessID, int newAccessID);
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/Optimizer/MultiStage.h]: done

File: src/dawn/Optimizer/PassDataLocalityMetric.cpp [hunk 25]
```diff
@@ -125,13 +128,23 @@ class ReadWriteCounter : public ASTVisitorForwarding {
...
+          (*iter).second.numReads++;
+        else
+          individualReadWrites_.emplace(AccessID, ReadWriteIDspecific(1, 0));
       }
       if(cache.getCacheIOPolicy() == Cache::flush ||
          cache.getCacheIOPolicy() == Cache::fill_and_flush) {
-        std::cout << "   Flush: " << getNameFromAccessID(AccessID) << std::endl;
+//        std::cout << "   Flush: " << getNameFromAccessID(AccessID) << std::endl;
         numWrites_++;
+        auto iter = individualReadWrites_.find(AccessID);
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/Optimizer/PassDataLocalityMetric.cpp]: the problem is that you actually need the iterator and do something different with it dependent on the situation. 
We could possibly write a function 
`std::pair<bool, typename T::iterator> accessIDisInContainer(int id, T container)`
but I don't think this really improves readabilty?

File: src/dawn/Optimizer/PassDataLocalityMetric.cpp [hunk 26]
```diff
@@ -218,23 +246,11 @@ class ReadWriteCounter : public ASTVisitorForwarding {
   }
 
   void visit(const std::shared_ptr<FieldAccessExpr>& expr) override { processReadAccess(expr); }
+  std::unordered_map<int, ReadWriteIDspecific> getIndividualReadWrites() const {
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/Optimizer/PassDataLocalityMetric.cpp]: oversight, fixed

File: src/dawn/Optimizer/PassDataLocalityMetric.h [hunk 27]
```diff
@@ -27,6 +28,16 @@ struct HardwareConfig {
   int TexCacheMaxFields = 3;
 };
 
+struct ReadWriteIDspecific {
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/Optimizer/PassDataLocalityMetric.h]: If you want to use emplace to emplace a key-value pair of int and struct in a map, is an explicit ctor not needed?

File: src/dawn/Optimizer/PassDataLocalityMetric.h [hunk 28]
```diff
@@ -41,6 +52,14 @@ class PassDataLocalityMetric : public Pass {
   HardwareConfig config_;
 };
 
+std::pair<int, int> computeReadWriteAccessesMetric(StencilInstantiation* instantiation,
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/Optimizer/PassDataLocalityMetric.h]: This was here previously... I think it returns total number of Reads and total number of Writes

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp (hunk 12 above)
Comment (twicki) [reply src/dawn/Optimizer/PassSetNonTempCaches.cpp]: sure

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp (hunk 1 above)
Comment (twicki) [reply src/dawn/Optimizer/PassSetNonTempCaches.cpp]: again, emplace does not work if we do not have this. Is there a better way?

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp (hunk 14 above)
Comment (twicki) [reply src/dawn/Optimizer/PassSetNonTempCaches.cpp]: fixed

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp (hunk 13 above)
Comment (twicki) [reply src/dawn/Optimizer/PassSetNonTempCaches.cpp]: fixed

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp (hunk 15 above)
Comment (twicki) [reply src/dawn/Optimizer/PassSetNonTempCaches.cpp]: sounds good.

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp (hunk 16 above)
Comment (twicki) [reply src/dawn/Optimizer/PassSetNonTempCaches.cpp]: We can call it `iter` since it actually is an iterator in the map that checks if we already checked this field

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp (hunk 17 above)
Comment (twicki) [reply src/dawn/Optimizer/PassSetNonTempCaches.cpp]: I was thinking of having them there for readability but can see the argument for removing them

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp (hunk 8 above)
Comment (twicki) [reply src/dawn/Optimizer/PassSetNonTempCaches.cpp]: we can do this

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp (hunk 9 above)
Comment (twicki) [reply src/dawn/Optimizer/PassSetNonTempCaches.cpp]: sure

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp (hunk 6 above)
Comment (twicki) [reply src/dawn/Optimizer/PassSetNonTempCaches.cpp]: after changing it I am also not super pleased with how auto looks. In either case this is a noisy part of the code as we manually create a stage in memory. Whatever works better

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp (hunk 5 above)
Comment (twicki) [reply src/dawn/Optimizer/PassSetNonTempCaches.cpp]: it actually is only 2 consecutive calls. Is it really worth it for that?

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp (hunk 20 above)
Comment (twicki) [reply src/dawn/Optimizer/PassSetNonTempCaches.cpp]: Should it be moved around though? Should this not in the end be a singleton that is in a place where it is globally accessible for all passes?

File: src/dawn/Optimizer/PassDataLocalityMetric.h (hunk 27 above)
[the remark this answers is not in the document]
Comment (thfabian) [reply src/dawn/Optimizer/PassDataLocalityMetric.h]: Possibly true, but you can always write `map.emplace(ReadWriteAccumulator{X,Y})` which is equally efficient.

[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/Optimizer/PassDataLocalityMetric.h]: then you would have to remove the default values. A trade off I am willing to do

File: src/dawn/Optimizer/PassDataLocalityMetric.cpp [hunk 29]
```diff
@@ -125,13 +128,23 @@ class ReadWriteCounter : public ASTVisitorForwarding {
     if(cache.getCacheType() == Cache::K) {
       if(cache.getCacheIOPolicy() == Cache::fill ||
          cache.getCacheIOPolicy() == Cache::fill_and_flush) {
-        std::cout << "   Fill: " << getNameFromAccessID(AccessID) << std::endl;
+//        std::cout << "   Fill: " << getNameFromAccessID(AccessID) << std::endl;
```
Comment (cosunae) [inline src/dawn/Optimizer/PassDataLocalityMetric.cpp]: this is dead code, can we remove it?

Comment (twicki) [reply src/dawn/Optimizer/PassDataLocalityMetric.cpp]: Well, the whole pass could be refactored. These were just cout's in the middle of the computation. We can go one of two ways: Either we let the pass return a struct that holds all the info that is cout'ed and store it somewere (probably in the same place where the hardware-config singleton should be) or we make it s.t the pass only keeps the data locally and add a proper if verbose-stmt to it.

For now I did not want to kill all the output that the pass had for good so I left in in this stage until the decision is made

File: src/dawn/Optimizer/PassDataLocalityMetric.cpp [hunk 30]
```diff
@@ -125,13 +128,23 @@ class ReadWriteCounter : public ASTVisitorForwarding {
...
         numReads_++;
+        auto iter = individualReadWrites_.find(AccessID);
```
Comment (cosunae) [inline src/dawn/Optimizer/PassDataLocalityMetric.cpp]: just wondering, would the one line
```
individualReadWrites_[AccessID].numReads++;
```
do the whole magic? individualReadWrites_[AccessID] would default construct an element if key is not found...

File: src/dawn/Optimizer/PassDataLocalityMetric.cpp [hunk 31]
```diff
@@ -125,13 +128,23 @@ class ReadWriteCounter : public ASTVisitorForwarding {
...
+        if(iter != individualReadWrites_.end())
+          (*iter).second.numReads++;
+        else
+          individualReadWrites_.emplace(AccessID, ReadWriteIDspecific(1, 0));
       }
       if(cache.getCacheIOPolicy() == Cache::flush ||
          cache.getCacheIOPolicy() == Cache::fill_and_flush) {
-        std::cout << "   Flush: " << getNameFromAccessID(AccessID) << std::endl;
+//        std::cout << "   Flush: " << getNameFromAccessID(AccessID) << std::endl;
```
Comment (cosunae) [inline src/dawn/Optimizer/PassDataLocalityMetric.cpp]: remove?

File: src/dawn/Optimizer/PassDataLocalityMetric.cpp (hunk 25 above)
[the remark this answers is not in the document]
Comment (cosunae) [reply src/dawn/Optimizer/PassDataLocalityMetric.cpp]: I made a same comment above on the reads...

File: src/dawn/Optimizer/PassDataLocalityMetric.cpp [hunk 32]
```diff
@@ -142,8 +155,13 @@ class ReadWriteCounter : public ASTVisitorForwarding {
 
     // Is field stored in cache?
     if(!multiStage_.getCaches().count(AccessID)) {
-      std::cout << "   Write: " << getNameFromAccessID(AccessID) << std::endl;
+//      std::cout << "   Write: " << getNameFromAccessID(AccessID) << std::endl;
```
Comment (cosunae) [inline src/dawn/Optimizer/PassDataLocalityMetric.cpp]: remove?

File: src/dawn/Optimizer/PassDataLocalityMetric.cpp (hunk 26 above)
[the remark this answers is not in the document]
Comment (cosunae) [reply src/dawn/Optimizer/PassDataLocalityMetric.cpp]: is not fixed yet, isnt? Actually the method is const, so shouldnt it return  a const & ?

File: src/dawn/Optimizer/PassDataLocalityMetric.cpp (hunk 30 above)
Comment (twicki) [reply src/dawn/Optimizer/PassDataLocalityMetric.cpp]: This is actually really smart :+1:

File: src/dawn/Optimizer/PassDataLocalityMetric.cpp [hunk 33]
```diff
@@ -286,6 +302,38 @@ computeReadWriteAccessesLowerBound(StencilInstantiation* instantiation,
...
+
+
+/// @brief Approximate the reads and writes accoding to our data locality metric
+std::pair<int, int> computeReadWriteAccessesMetric(StencilInstantiation* instantiation,
+                                                   const MultiStage& multiStage,
+                                                   const HardwareConfig& config) {
+  ReadWriteCounter readWriteCounter(instantiation, multiStage, config);
+
+  for(const auto& stage : multiStage.getStages())
+    for(const auto& DoMethod : stage->getDoMethods())
```
Comment (cosunae) [inline src/dawn/Optimizer/PassDataLocalityMetric.cpp]: question about convention: do we have a convention for if, ...scopes? I would always  {} (as opposed to single line) since it can avoid mistakes when we add more statements. But if we have a convention for single line without {} already is fine.... as far as we are consistent

File: src/dawn/Optimizer/PassDataLocalityMetric.h (hunk 28 above)
[the remark this answers is not in the document]
Comment (cosunae) [reply src/dawn/Optimizer/PassDataLocalityMetric.h]: sorry, I didnt get it. Why not using here ReadWriteAccumulator better than std::pair? like that we know what each member of the pair means, so is more readable.
Also why do we pass StencilInstantiation* instead of reference of shared_ptr ?
And my final comment here, can we add documentation (doxygen) what the function does?

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 34]
```diff
@@ -0,0 +1,322 @@
...
+
+struct NameToImprovementMetric {
+  std::string name;
+  Cache cache;
+  int dataLocalityImprovement;
+};
+
+class GlobalFieldCacher {
+public:
+  /// @brief The GlobalFieldCacher class handles the caching for a given Multistage
```
Comment (cosunae) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: shouldnt this documentation be above the class declaration (rather than ctr)?

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 35]
```diff
@@ -0,0 +1,322 @@
...
+
+  /// @brief entry method for the pass: processes a given multistage and applies all changes
+  /// required
+  void process() {
+    computeOptimalFields();
+    addFillerStages();
+    addFlushStages();
+  }
+
+  const std::vector<NameToImprovementMetric>& getOriginalNameToCache() {
```
Comment (cosunae) [inline src/dawn/Optimizer/PassSetNonTempCaches.cpp]: should the method be const?

File: src/dawn/Optimizer/PassDataLocalityMetric.cpp (hunk 33 above)
Comment (thfabian) [reply src/dawn/Optimizer/PassDataLocalityMetric.cpp]: I never use the {} for a single line (I think it is also stated by the LLVM coding style). Imho it just pollutes the code with unnecessary braces.

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp (hunk 34 above)
Comment (thfabian) [reply src/dawn/Optimizer/PassSetNonTempCaches.cpp]: The `@brief` statement yes, but not the `@param`. I think I wasn't entirely clear in my review on this.

File: src/dawn/Optimizer/PassDataLocalityMetric.h (hunk 27 above)
[the remark this answers is not in the document]
Comment (thfabian) [reply src/dawn/Optimizer/PassDataLocalityMetric.h]: I don't get it, why would you need to remove the default values? 
```
ReadWriteAccumulator foo{}; // <-- calls default constructor
ReadWriteAccumulator bar{5, 5}; // <-- performs aggregate initialization
```

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp (hunk 20 above)
Comment (thfabian) [reply src/dawn/Optimizer/PassSetNonTempCaches.cpp]: Singletons in library is usually a bad design, I would argue you have to pass it around. You can also simply attach it to the `StencilInstantiation` i.e ISIR

File: src/dawn/Optimizer/PassDataLocalityMetric.h (hunk 27 above)
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/Optimizer/PassDataLocalityMetric.h]: you cannot write 

```
struct foo{
int i = 1;
int j = 2;
};

int main(){
 foo{2,3};
}
```

It was possible to work around that by carlos' fix, so this is not necessary anymore but just for the sake of completeness

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp (hunk 20 above)
Comment (twicki) [reply src/dawn/Optimizer/PassSetNonTempCaches.cpp]: ISIR seems like an ok place to put it. This allows for a very simple pass to set it properly. I see two problems with this though: 
- The first one is that we add an additional field to the ISIR which is not really an ISIR feature?
- This is not something that is not really dependent on a Stencil/Stencil-Function but is the same thing on a global scope. 

I'd suggest to add it to the optimizer.

Thoughts?
